[15.0][MIG] base_tier_validation_waiting

bosd opened the pull request:
Depends on: 
#732 Do not sent double messages on the first validation sequence

BT-jcolmeiro committed:
[ADD] base_tier_validation_waiting

Commit:
[UPD] Update base_tier_validation_waiting.pot

OCA-git-bot committed:
[UPD] README.rst

bosd committed:
[MIG] base_tier_validation_waiting: Migration to 15.0

bosd committed:
[IMP] base_tier_validation_waiting: pre-commit execution

bosd commented:
/ocabot migration base_tier_validation_waiting

simahawk commented:
/ocabot migration base_tier_validation_waiting

simahawk commented on base_tier_validation_waiting/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,2 @@
+This module extends base_tier_validation to add an initial **"waiting"** status to tier reviews.
+It also auto moves next review to pending status when needed and handles notification if set.
```
can you drop empty readme files pls?

simahawk commented on base_tier_validation_waiting/tests/test_tier_validation.py:
```diff
@@ -0,0 +1,101 @@
+    @classmethod
+    def setUpClass(cls):
+        super(TierTierValidation, cls).setUpClass()
+
```
can you set tracking_disable + no_reset_password context keys pls?

simahawk replied:
I meant class wise ;)
```suggestion
        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```

simahawk reviewed (commented):
LG minor comments

bosd commented:
@cwjoalder Can you please review?

bosd commented:
@simahawk Can you review the changes? :pray:

michaelslade54 commented on base_tier_validation_waiting/models/tier_validation.py:
```diff
@@ -0,0 +1,50 @@
+        self.ensure_one()
+        res = super(TierValidation, self)._validate_tier(tiers)
+        tier_reviews = tiers or self.review_ids
+        waiting_reviews = tier_reviews.filtered(
+            lambda r: r.status == "waiting"
+            and (self.env.user in r.reviewer_ids)
+            and r.approve_sequence_bypass
```
It appears this migration relies on PR #656 or am I missing something?

bosd replied:
Great Spot!! Seems it was missing in the migration.

bosd replied:
That pr is merged, so we can continue here.

bosd commented:
Force pushed to recreate runboat

bosd commented:
@Dranyel-Bosd Can you please review?

Dranyel-Bosd reviewed (approved):
looks so good to me :+1: :sparkler:

simahawk commented on base_tier_validation_waiting/tests/test_tier_validation.py:
```diff
@@ -0,0 +1,103 @@
+
+        cls.loader.update_registry((TierValidationTester,))
+        cls.test_model = cls.env[TierValidationTester._name].with_context(
+            tracking_disable=True, no_reset_password=True
```
see above

simahawk commented on base_tier_validation_waiting/models/tier_review.py:
```diff
@@ -0,0 +1,45 @@
+
+    def _notify_pending_status(self, review_ids):
+        """Method to call and reuse abstract notification method"""
+        resource = self.env[self.model].browse(self.res_id)
```
this is not tested, any chance?

bosd replied:
I've added tests on the notifications

simahawk commented on base_tier_validation_waiting/models/tier_review.py:
```diff
@@ -0,0 +1,45 @@
+                if record.status != "waiting":
+                    continue
+                # if approve by sequence, check sequence has been reached
+                if record.approve_sequence:
```
I'm not sure I get this IF block: in any case you set "pending" -> can't you have 1 single check?

simahawk commented on base_tier_validation_waiting/models/tier_review.py:
```diff
@@ -0,0 +1,45 @@
+                else:
+                    record.status = "pending"
+                # notify if state has changed
+                if record.status == "pending":
```
IMO this can should be done on inverse method for `status` field

simahawk commented on base_tier_validation_waiting/models/tier_review.py:
```diff
@@ -0,0 +1,45 @@
+                if record.status == "pending":
+                    if record.definition_id.notify_on_pending:
+                        record._notify_pending_status(record)
+        return super(TierReview, self)._compute_can_review()
```
can't we call super w/ a subset of the records?

simahawk commented on base_tier_validation_waiting/models/tier_validation.py:
```diff
@@ -0,0 +1,50 @@
+                    lambda r: r.definition_id.notify_on_pending and r.res_id == rec.id
+                ).mapped("reviewer_ids")
+                # Subscribe reviewers and notify
+                getattr(rec.sudo(), subscribe)(
```
why calling sudo every time if you can do `for rec in self.sudo()`?

simahawk replied:
Also, is sudo really needed?

bosd replied:
Apparantly it is: https://github.com/OCA/server-ux/pull/359#issue-936237493

simahawk commented on base_tier_validation_waiting/models/tier_validation.py:
```diff
@@ -0,0 +1,50 @@
+                )
+
+    def _validate_tier(self, tiers=False):
+        """extend base tier validation method to allow bypass of waiting
+        sequences as pending state is hardcoded"""
```
I'd drop this docstring and move the custom logic into `_validate_tier_waiting_reviews`

bosd committed:
[IMP] base_tier_validation_waiting test

Co-authored-by: Simone Orsi <simahawk@users.noreply.github.com>

bosd committed:
[IMP] base_tier_validation_waiting: validation state w/o sequence

bosd committed:
[IMP] base_tier_validation_waiting: code refactor

bosd committed:
[IMP] base_tier_validation_waiting: test notifications

bosd committed:
[IMP] base_tier_validation: update tests to support messages

bosd committed:
[IMP] base_tier_validation_forward: update tests to support messages

bosd committed:
[IMP] base_tier_validation_formula: update tests to support messages

bosd committed:
[IMP] base_tier_validation: use sudo to post message

bosd commented:
@simahawk Thanks for your review!

I'm working my way towards more coverage and green tests :grin: 
Adding coverage by testing the notifications.
But it's breaking/affecting tests of other modules.
Added the fixes for those in this PR. Is that allowed/correct OCA procedure?

I've watched your talks about the unittests. 
Still I'm struggling with getting how to get the tier_validation_server_action fixed.

This module is altering the behaviour by adding the state waiting, which results in breaking the tests of server action.
When this module is installed, an alternative version of tests should be run by that module. Is that correct?
Can you explain how to do that?  :blush: :pray: 

```

===================================================================== 
2023-10-23 07:04:27,754 1037 ERROR odoo odoo.addons.base_tier_validation_server_action.tests.test_tier_validation: FAIL: TierTierValidation.test_1_auto_validation
Traceback (most recent call last):
  File "/__w/server-ux/server-ux/base_tier_validation_server_action/tests/test_tier_validation.py", line 100, in test_1_auto_validation
    self.assertEqual(
AssertionError: Lists differ: ['waiting', 'waiting', 'waiting'] != ['pending', 'pending', 'pending']

First differing element 0:
'waiting'
'pending'

- ['waiting', 'waiting', 'waiting']
+ ['pending', 'pending', 'pending']
 
 ```

DavidJForgeFlow commented:
```
from odoo import api, fields, models
from odoo.tools import config


class TierReview(models.Model):
    _inherit = "tier.review"


    def _default_status(self):
        if not config["test_enable"] or config["test_enable"] and self.env.context.get("_testing_base_tier_validationwaiting"):
            self.status = "waiting"
        else:
            self.status = "pending"
        
    status = fields.Selection(
        selection_add=[("waiting", "Waiting")],
        default="_default_status",
        ondelete={"waiting": "set default"},
    )
```
Maybe this helps to work with the other module

bosd committed:
[IMP] base_tier_validation_waiting: Improve test compatibility

without this the tests of some other modules might fail, e.g. base_tier_validation_server_action.

The introduced context prevents writing/patching unit tests of other base_tier_validation_* modules.

Dranyel-Bosd reviewed (approved):
LGTM :+1:

bosd commented:
Dropped the dependency on 732. Now we finally have green tests  :heavy_check_mark: 
:sparkles: :tada:

simahawk commented on base_tier_validation_waiting/models/tier_review.py:
```diff
@@ -2,14 +2,25 @@
 
+    def _default_status(self):
+        if (
+            not config["test_enable"]
```
do we really need this specific check for tests?

bosd replied:
Yes, without this commit, it would mean that we need to patch a lot of depending modules, to fix their tests.

Another solution I've evaluated is using seperate tests. That might solve the problem with the CI tests here on github.
But it would simply transfer the problem to the persons running an custom CI like odoo.sh.

The method here is not really clean. But it's the least dirty solution.
(Discussed this test issue with a couple of oca contributors at the OXP)
Maybe you have another insight to solve this problem?

simahawk commented on base_tier_validation_waiting/models/tier_review.py:
```diff
@@ -2,14 +2,25 @@
     status = fields.Selection(
         selection_add=[("waiting", "Waiting")],
-        default="waiting",
+        default=_default_status,
```
use lambda to not reference the method object directly
